add quantization debug util to pretty print FX graphs

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43928 fx quant: add docblocks to _find_matches and _find_quants
* #43927 fx quant: clarify state in Quantizer object
* #43923 fx quant: make load_arg function more clear
* #43914 fx quant: rename matches -> is_match
* **#43910 add quantization debug util to pretty print FX graphs**

Summary:

Adds a debug function to get a representation of all nodes in the
graph, such as

```
name          op      target         args               kwargs
x             plchdr  x              ()                 {}
linear_weight gt_prm  linear.weight  ()                 {}
add_1         cl_fun  <bi_fun add>   (x, linear_weight) {}
linear_1      cl_mod  linear         (add_1,)           {}
relu_1        cl_meth relu           (linear_1,)        {}
sum_1         cl_fun  <bi_meth sum>  (relu_1,)          {'dim': -1}
topk_1        cl_fun  <bi_meth topk> (sum_1, 3)         {}
```

using only Python STL. This is useful for printing internal state of
graphs when working on FX code.

Has some on-by-default logic to shorten things so that node reprs for
toy models and unit tests fit into 80 chars.

Flexible on function name and location, I care more that this is
accessible from both inside PT as well as from debug scripts which
are not checked in.

Test Plan: see
https://gist.github.com/vkuzo/ed0a50e5d6dc7442668b03bb417bd603 for
example usage

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23435029](https://our.internmc.facebook.com/intern/diff/D23435029)

vkuzo committed:
add quantization debug util to pretty print FX graphs

Summary:

Adds a debug function to get a representation of all nodes in the
graph, such as

```
name          op      target         args               kwargs
x             plchdr  x              ()                 {}
linear_weight gt_prm  linear.weight  ()                 {}
add_1         cl_fun  <bi_fun add>   (x, linear_weight) {}
linear_1      cl_mod  linear         (add_1,)           {}
relu_1        cl_meth relu           (linear_1,)        {}
sum_1         cl_fun  <bi_meth sum>  (relu_1,)          {'dim': -1}
topk_1        cl_fun  <bi_meth topk> (sum_1, 3)         {}
```

using only Python STL. This is useful for printing internal state of
graphs when working on FX code.

Has some on-by-default logic to shorten things so that node reprs for
toy models and unit tests fit into 80 chars.

Flexible on function name and location, I care more that this is
accessible from both inside PT as well as from debug scripts which
are not checked in.

Test Plan: see
https://gist.github.com/vkuzo/ed0a50e5d6dc7442668b03bb417bd603 for
example usage

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/fx/utils.py:
```diff
@@ -1,7 +1,76 @@
         return r[0], r[1]
+
+
+def graph_pretty_str(g, shorten=True) -> str:
```
btw, james has an example to print a table in his slides:

```
class MyModule(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.param = torch.nn.Parameter(torch.rand(3, 4))
        self.linear = torch.nn.Linear(4, 5)
        
    def forward(self, x):
        return torch.relu(self.linear(x + self.param))
    
m = MyModule()
gm = GraphModule(m)

from tabulate import tabulate
node_specs = [[n.op, n.target, n.args, n.kwargs] for n in gm.graph.nodes]
print(tabulate(node_specs, headers=['opcode', 'target', 'args', 'kwargs']))
```
would be good if we can reuse the code that generate a table

vkuzo replied:
agreed.  One issue is that `tabulate` is not in Python STL and isn't currently required by pytorch.

jerryzh168 replied:
yeah sounds good. we can change this if torch.fx decide to add tabulate as a dependency

jerryzh168 reviewed (approved):
Looks good, thanks

jerryzh168 commented:
I think it would also be helpful to provide a function to print node

vkuzo committed:
Update on "add quantization debug util to pretty print FX graphs"


Summary:

Adds a debug function to get a representation of all nodes in the
graph, such as

```
name          op      target         args               kwargs
x             plchdr  x              ()                 {}
linear_weight gt_prm  linear.weight  ()                 {}
add_1         cl_fun  <bi_fun add>   (x, linear_weight) {}
linear_1      cl_mod  linear         (add_1,)           {}
relu_1        cl_meth relu           (linear_1,)        {}
sum_1         cl_fun  <bi_meth sum>  (relu_1,)          {'dim': -1}
topk_1        cl_fun  <bi_meth topk> (sum_1, 3)         {}
```

using only Python STL. This is useful for printing internal state of
graphs when working on FX code.

Has some on-by-default logic to shorten things so that node reprs for
toy models and unit tests fit into 80 chars.

Flexible on function name and location, I care more that this is
accessible from both inside PT as well as from debug scripts which
are not checked in.

Test Plan: see
https://gist.github.com/vkuzo/ed0a50e5d6dc7442668b03bb417bd603 for
example usage

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23435029](https://our.internmc.facebook.com/intern/diff/D23435029)

[ghstack-poisoned]

vkuzo committed:
Update on "add quantization debug util to pretty print FX graphs"


Summary:

Adds a debug function to get a representation of all nodes in the
graph, such as

```
name          op      target         args               kwargs
x             plchdr  x              ()                 {}
linear_weight gt_prm  linear.weight  ()                 {}
add_1         cl_fun  <bi_fun add>   (x, linear_weight) {}
linear_1      cl_mod  linear         (add_1,)           {}
relu_1        cl_meth relu           (linear_1,)        {}
sum_1         cl_fun  <bi_meth sum>  (relu_1,)          {'dim': -1}
topk_1        cl_fun  <bi_meth topk> (sum_1, 3)         {}
```

using only Python STL. This is useful for printing internal state of
graphs when working on FX code.

Has some on-by-default logic to shorten things so that node reprs for
toy models and unit tests fit into 80 chars.

Flexible on function name and location, I care more that this is
accessible from both inside PT as well as from debug scripts which
are not checked in.

Test Plan: see
https://gist.github.com/vkuzo/ed0a50e5d6dc7442668b03bb417bd603 for
example usage

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23435029](https://our.internmc.facebook.com/intern/diff/D23435029)

[ghstack-poisoned]
